src/Event/BatchActionEvent.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Event;
+
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Sonata\AdminBundle\Datagrid\ProxyQueryInterface;
+use Symfony\Contracts\EventDispatcher\Event;
+
+/**
+ * This event is sent by hook:
+ *   - preBatchAction.
+ *
+ * You can register the listener to the event dispatcher by using:
+ *   - sonata.admin.event.batch_action.pre_batch_action)
+ *   - sonata.admin.event.batch_action.[admin_code].pre_batch_action  (not implemented yet)
+ *
+ * @author Jochem Klaver <info@7ochem.nl>
+ *
+ * @phpstan-template T of object
+ */
+final class BatchActionEvent extends Event
+{
+    public const TYPE_PRE_BATCH_ACTION = 'pre_batch_action';
+    public const TYPE_POST_BATCH_ACTION = 'post_batch_action'; // not implemented yet
+
+    /**
+     * @var AdminInterface<object>
+     * @phpstan-var AdminInterface<T>
+     */
+    private $admin;
+
+    /**
+     * @var string
+     * @phpstan-var self::TYPE_*
+     */
+    private $type;
+
+    /**
+     * @var string
+     */
+    private $actionName;
+
+    /**
+     * @var ProxyQueryInterface
+     */
+    private $proxyQuery;
+
+    /**
+     * @var mixed[]
+     */
+    private $idx;
+
+    /**
+     * @var bool
+     */
+    private $allElements;
+
+    /**
+     * @param mixed[] $idx
+     * @phpstan-param AdminInterface<T> $admin
+     * @phpstan-param self::TYPE_* $type
+     */
+    public function __construct(AdminInterface $admin, string $type, string $actionName, ProxyQueryInterface $proxyQuery, array &$idx, bool $allElements = false)
+    {
+        $this->admin = $admin;
+        $this->type = $type;
+        $this->actionName = $actionName;
+        $this->proxyQuery = $proxyQuery;
+        $this->idx = &$idx;
+        $this->allElements = $allElements;
+    }
+
+    /**
+     * @phpstan-return AdminInterface<T>
+     */
+    public function getAdmin(): AdminInterface
+    {
+        return $this->admin;
+    }
+
+    /**
+     * @phpstan-return self::TYPE_*
+     */
+    public function getType(): string
+    {
+        return $this->type;
+    }
+
+    public function getActionName(): string
+    {
+        return $this->actionName;
+    }
+
+    public function getProxyQuery(): ProxyQueryInterface
+    {
+        return $this->proxyQuery;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function &getIdx(): array
```

Comment:
```suggestion
    public function getIdx(): array
```
?

Reply:
The array of IDs selected for the batch actions is passed by reference everywhere. Also to the `AbstractAdmin::preBatchAction()` for example. This enables you to manipulate the IDs. So here it should also be returned by reference, hence the `&` in front of the function name.

This enables you to manipulate the array of IDs in your event listener/subscriber:

```php
$idx = &$event->getIdx();
$idx[] = 123;
```

Above will result in the batch action also being executed on object with ID 123. Of course this is an example and normally you'll have some business logic in there. And you could also remove IDs of course.

This can already be done in an Admin Extension's `preBatchAction()` method.

Reply:
Didn't know this syntax. Thanks for the explanation.

Reply:
Wouldn't be better to have getters and setters?

Reply:
Like:
```php
$idx = $event->getIdx();
$idx[] = 123;
$event->setIdx($idx);
```
That would be possible. I tried to be consistent with the concept of passing by reference that is already in the existing code for `preBatchAction`